Mandate QUIC as primary transport (#5330)
ethereum/consensus-specs: specs/phase0/p2p-interface.md

simultaneous transports and endpoints transparently), we hereby define a profile
for basic interoperability.

All implementations MUST support the QUIC (UDP) libp2p transport and SHOULD
support the TCP libp2p transport as a fallback. Any supported transport MUST be
enabled for both dialing and listening (i.e. outbound and inbound connections).
The libp2p QUIC (UDP) and TCP transports support listening on IPv4 and IPv6

> addresses (and on multiple simultaneously).

Reviewer: ```suggestion
All implementations MUST support the QUIC (UDP) libp2p transport and the TCP libp2p transport. Any supported transport MUST be
enabled for both dialing and listening (i.e. outbound and inbound connections).
The libp2p QUIC (UDP) and TCP transports support listening on IPv4 and IPv6
addresses (and on multiple simultaneously).
```

QUIC is the primary transport. When a peer is reachable over more than one
transport (for example, it advertises both `quic` and `tcp` endpoints in its
ENR), clients SHOULD attempt to dial it over QUIC first and fall back to TCP

> only if a QUIC connection cannot be established.

Reviewer: ```suggestion
QUIC is the primary transport. When a peer is reachable over both TCP and QUIC, clients SHOULD prioritise peer's QUIC addresses.
```

The original phrasing is too restrictive. The way to do this is some happy eyeballs like system where you delay the tcp dial for some time(generally estimated 1RTT). This is much lower than what your transport stack would setup as a connection failure timeout. 

Here's the go-libp2p logic: https://github.com/libp2p/go-libp2p/blob/master/p2p/net/swarm/dial_ranker.go#L54

### Encryption and identification

The QUIC transport is natively secured using TLS 1.3, so no separate libp2p

> secure channel handshake is layered on top of it.

Author: ```suggestion
The QUIC transport is secured using TLS 1.3, configured according to the [libp2p TLS specification](https://github.com/libp2p/specs/blob/master/tls/tls.md).
```

### Encryption and identification

The QUIC transport is natively secured using TLS 1.3, so no separate libp2p

> secure channel handshake is layered on top of it.

Reviewer: we need to specify that we use the libp2p tls handshake to verify peer id right?

[yamux](https://github.com/libp2p/specs/blob/master/yamux/README.md). Their
protocol IDs are, respectively: `/mplex/6.7.0` and `/yamux/1.0.0`.

Because the primary QUIC transport multiplexes natively, a stream multiplexer is
only required for the TCP fallback transport. Clients that support the TCP

> transport MAY support [mplex](https://github.com/libp2p/specs/tree/master/mplex)

Reviewer 2: ```suggestion
transport MUST support [mplex](https://github.com/libp2p/specs/tree/master/mplex)
```

yamux never made it to the required list of protocols so it's not implemented in all clients.

Reviewer 3: +1 on this proposal. 
As soon as the TCP + Noise is assumed to be a fallback protocol, I would prefer to fallback to a simple, reliable and battle tested muxer  which is mplex. 
Yamux is definitely more advanced but it's pretty complex because of this. 
Not sure about other implementations but in JVM libp2p it's still in alpha stage, and I would be quite uncomfortable to leave it as a part of the fallback stack.

Author: Updated.

that powers much of the Internet as we know it today. HTTP/1.1 and HTTP/2 run
atop TCP.

QUIC is a UDP-based transport standardized by the IETF in
[RFC 9000](https://www.rfc-editor.org/rfc/rfc9000), having emerged from Google’s
SPDY experiment. It’s reliable, ordered, multiplexed, natively secure (TLS 1.3),
reduces latency vs. TCP, and offers stream-level and connection-level congestion
control (thus removing head-of-line blocking), 0-RTT connection establishment,
and endpoint migration, amongst other features. UDP also has better NAT
traversal properties than TCP—something we desperately pursue in peer-to-peer
networks. These characteristics are why QUIC is the primary transport for the

> consensus-layer network.

Reviewer: ```suggestion
QUIC is a UDP-based transport standardized by the IETF in
[RFC 9000](https://www.rfc-editor.org/rfc/rfc9000). It’s reliable, ordered, multiplexed, natively secure (TLS 1.3),
reduces connection establishmen latency vs. TCP, offers stream-level flow control(thus removing head-of-line blocking), 0-RTT connection establishment,
and endpoint migration, amongst other features. UDP also has better NAT
traversal properties than TCP—something we desperately pursue in peer-to-peer
networks. These characteristics are why QUIC is the primary transport for the
consensus-layer network.
```

QUIC is also the underlying protocol for HTTP/3. This has the potential to award

> us censorship resistance via deep packet inspection for free. Provided that we

Reviewer: ```suggestion
QUIC is also the underlying protocol for HTTP/3. This has the potential to award
us censorship resistance against deep packet inspection for free. Provided that we
```

Reviewer: Not sure this is true though. We'd need ECH to make this work otherwise the ALPN is still sent in cleartext.

Author: This was was just inherited language, we can delete this sentence if you want.
